### ISSUE TITLE
useAnimatedProps is not working

### ISSUE THREAD

**Issue by kyle-ssg** [software-mansion/react-native-reanimated#1128]
## Description

From what I can see there are a few issues here, but basically I haven't been able to get useAnimatedProps to work in any usecase.

# Replication
https://github.com/kyle-ssg/reanimated2-experiments/blob/master/App/components/AnimatedOnOffImage.tsx

Given a simple example there's perhaps maybe a clue of What's happening, when the image source uri property is adjusted from the previous I see.
![image](https://user-images.githubusercontent.com/8608314/90313230-c9652880-df02-11ea-8a12-1db6c037dbe4.png)

<img src="http://g.recordit.co/RQ69YL25tl.gif"/>

Then also given the following code:

https://github.com/kyle-ssg/reanimated2-experiments/blob/master/App/components/AnimatedBlur.tsx#L43
<img src="http://g.recordit.co/pd7PmGeVrY.gif"/>

## Package versions

```
    "react": "16.13.1",
    "react-native": "0.63.1",
    "react-native-reanimated": "2.0.0-alpha.5",
```

**Comment by terrysahaidak** [comment]
This is more like a problem with BlurView itself. Are you able to reproduce this with Reanimated 1?

**Comment by kyle-ssg** [comment]
@terrysahaidak Hmm I thought so, but then the simple case with just an Image?

**Comment by kyle-ssg** [comment]
I've just tried with the TextInput example code 

<img src="https://recordit.co/tXkyg3gkZF.gif"/>

**Comment by terrysahaidak** [comment]
For text input, you need to pass initial text as a `value` and then animating text as a `text` property.

**Comment by kyle-ssg** [comment]
Hmm ok, I guess I don't really understand what use cases useAnimatedProps work for then. Are there rules that determine what properties are supported ? Every property I've tried so far either doesn't work or throws an error.

**Comment by terrysahaidak** [comment]
First of all, I recommend you to try to use the prop you're testing with Reanimated 1 first. If this doesn't work – it won't work with Reanimated 2.

Unfortunately, not all properties can be animated. But if something works for 1 then it should work the same way for 2.

Would be cool to have a list of properties that work for 1 and don't for 2.

**Comment by kyle-ssg** [comment]
Ok fair enough. I haven't actually found one that works in v2 at the moment, they all seem to initialise fine but then not update but I'll try a few out on v1 and see what I find and post here.

**Comment by terrysahaidak** [comment]
Cool, thank you, looking forward to it

**Comment by kyle-ssg** [comment]
Could you point me in the direction of how this is done in v1? I'm not sure if it's me being dumb but I can't see anything like this mentioned in the old docs.

**Comment by terrysahaidak** [comment]
With Reanimated 1 you need to createAnimatedComponent and pass Animated.Value as a prop you want to animate. Check out Retext from react-native-redash.

**Comment by kyle-ssg** [comment]
Ok I'm starting to understand this a bit better, you were right about the blur view, I've added a [PR to the lib](https://github.com/react-native-community/react-native-blur/pull/392) that fixes iOS and Android to support reanimated 2. There does seem to be 1 key rule in that the component in question needs to expose a setNativeProps function.

<img 
align="left"
height="400"
src="https://camo.githubusercontent.com/8381e59577745fa936712ea7df7be8a859821ed3/687474703a2f2f672e7265636f726469742e636f2f424a6d74736b4a6f4e572e676966"/>

<img 
align="left"
height="400"
src="https://camo.githubusercontent.com/756ed76f7e151172542b899c211afd18356516ca/68747470733a2f2f7265636f726469742e636f2f786a76447939547074342e676966"/>

I'm going to see what's going on with images.

**Comment by terrysahaidak** [comment]
The library should provide setNativeProps natively - I mean the underlying native view should provide it. Because Reanimated doesn't use js version of setNativeProps.

**Comment by kyle-ssg** [comment]
Hmm, odd, it hit the js version when I added it. Without adding this I simply hit a load of undefined errors.

https://github.com/kyle-ssg/react-native-blur/blob/master/src/BlurView.android.js#L66

**Comment by terrysahaidak** [comment]
Yeah, you're right too. It sets the initial state of the view I think - that's why it needs it.

**Comment by kyle-ssg** [comment]
Hey @cltsang unfortunately I couldn't find a way to achieve this

**Comment by cltsang** [comment]
Thanks for the quick reply. That's unfortunate.

**Comment by kyle-ssg** [comment]
I think the issue is quite low level and may be down to the React Native side, might be worth creating a new issue for maintainers here to look at.

**Comment by cltsang** [comment]
For now, to get the job done, I choose to cross the JS bridge to achieve the same visual effect
So to use your example, instead of
``` js
  const animatedProps = useAnimatedProps(() => {
    return {
      source: {
        uri:
          animatedValue.value >= 0.5
            ? 'https://via.placeholder.com/350x150'
            : 'https://via.placeholder.com/100x100',
      },
    };
  });
```
It would be
``` js
const [imageUri, setImageUri] = useState('https://via.placeholder.com/350x150'); // assuming its initial value is >= 0.5
useAnimatedReaction(
  () => {
    return animatedValue.value;
  }, (result) => {
    runOnJS(setImageUri)(result  >= 0.5 ? 'https://via.placeholder.com/350x150' : 'https://via.placeholder.com/100x100');
  }
);
```

And apply the uri on regular `Image` rather than `AnimatedImage`.

If you have just a few image sources that aren't demanding, it makes no difference visually. Frame rate stays at 60. But of course, proper clean code would be so much nicer.

**Comment by kyle-ssg** [comment]
Yep nice that makes sense, I guess yeah it'll stay at 60 so long as there's no work blocking JS thread

**Comment by kyle-ssg** [comment]
@cltsang  You could probably make that a bit more efficient by only calling runOnJS if the new image URI will be different from previous

```
  }, (result,previousResult) => { if(result!==previousResult)

```

**Comment by cltsang** [comment]
Right! Thank you.

**Comment by kyle-ssg** [comment]
Maybe this 
```
const [imageUri, setImageUri] = useState('https://via.placeholder.com/350x150'); // assuming its initial value is >= 0.5
useAnimatedReaction(
  () => {
    return animatedValue.value  >= 0.5 ? 'https://via.placeholder.com/350x150' : 'https://via.placeholder.com/100x100'
  }, (result,previous) => {
   if(result!==previous) {
    runOnJS(setImageUri)();
    }
  }
);
```
